### DISCUSSION TITLE
Is there a way to get the entity's position right after creating it?

### DISCUSSION THREAD

**Discussion by marcusx2** [AR-js-org/AR.js#497] [Help]
@nickw1 If I add an entity to the DOM at runtime, I can't get its position without first calling setTimeout to wait 1 frame.

```
entity = document.createElement('a-entity');
entity.setAttribute('gps-new-entity-place', {latitude: somevalue, longitude: somevalue});
document.querySelector('a-scene').appendChild(entity);
var floatArray = [entity.object3D.position.x, entity.object3D.position.y, entity.object3D.position.z];//returns [0, 0, 0]
```

```
entity = document.createElement('a-entity');
entity.setAttribute('gps-new-entity-place', {latitude: somevalue, longitude: somevalue});
document.querySelector('a-scene').appendChild(entity);
setTimeout(() => {
    var floatArray = [entity.object3D.position.x, entity.object3D.position.y, entity.object3D.position.z];//returns correct values. Needed to wait a frame
}, 0);//wait a frame! Position values are not immediately available...
```
My use case here, as I explained on another github issue, is that I am not actually using the entities themselves on the DOM, but just retrieving their transform information(position) to pass to PlayCanvas for example for positioning. I'm just using ARJS' api to retrieve the information I need to place AR content in another context.

Is there a way that I can do this without waiting a frame, synchronously? I don't mind if I need extra code to calculate the position myself. Thanks!

**Comment by nickw1** [comment]
Hello @marcusx2 you can do this with the camera component's `latLonToWorld()` method, e.g.

```
const camera = document.querySelector("[gps-new-camera]");
const [worldX, worldZ] = camera.components["gps-new-camera"].latLonToWorld(lat,lon);
```

**Reply by marcusx2** [reply]
@nickw1 This does work, however how do I get the `worldY` value in the same fashion? It's usually something small like 1.6. Otherwise I'll still need to use the setTimeout trick. Thanks!

EDIT

I found this code snippet `this.threeLoc.setElevation(this.data.simulateAltitude+1.6)` on the `aframe-ar` file. It seems to indicate that the worldY will always be the value of `simulateAltitude + 1.6`. Is that it? Or can the entity have a position component to further offset its position after the latitude/longitude is applied? In this case the entity's y position would be `simulateAltitude + 1.6 + positionY.`

EDIT

Ok so sorry, actually it's the camera's y position that is simulateAltitude + 1.6. An entity's y position is always 0 unless I set the position component, then it will read the y for the height. x and z is latitude and longitude and the position component doesn't affect those from my tests. Did I get it all correct this time? Thanks!

**Reply by nickw1** [reply]
@marcusx2 yes, that's about right. You can simulate the altitude with `simulateAltitude` and that sets the `y` coordinate of the camera to the simulated altitude + 1.6 (the 1.6 is an approximation to the height of a person in metres, so it's supposed to represent eye level).

You can also set the `y` coordinate of the camera at any other time using standard A-Frame API techniques. By default it will be 1.6, as `simulateAltitude` is 0 by default.

EDIT: note that `x` and `z` are not longitude and latitude, but they correspond to longitude and latitude. `x` and `z` are values obtained via the Spherical Mercator projection, and are approximately equivalent to, but not exactly, metres. (In particular note that one unit in the `x` direction will be a different distance depending on your latitude, but in tests at locations between 40 and 51 North, it's good enough. Nearer the equator it will be even better).

**Reply by marcusx2** [reply]
> You can also set the y coordinate of the camera at any other time using standard A-Frame API techniques. By default it will be 1.6, as simulateAltitude is 0 by default.

Setting the position component doesn't set the y coordinate for the camera. Example if I do `position='0 10 0'` on the camera, the y position will still be `simulateAltitude + 1.6`. The position component is completely ignored on the camera.

**Reply by nickw1** [reply]
It will be ignored if you statically set it, as the `simulateAltitude` overrides the `y` of the position. However if you later change the y coordinate dynamically from JavaScript, it will update.
